### PR TITLE
SOF-6588 feat: add getConsistencyChecks

### PR DESCRIPTION
**Pull request by VsevolodX**
This is the draft approach to test the waters. Checks for atom feasibility (verification? validation?) should happen in Made.Material. The generation of check result message or warning should be done in a separate class.

### CODE REVIEW

File: src/material.js [hunk 1]
```diff
@@ -319,4 +319,31 @@ export class Material extends HasMetadataNamedDefaultableInMemoryEntity {
...
+        overlappingGroups.forEach((group) => {
+            group.forEach((atom) => {
+                const otherAtoms = group
+                    .filter((other) => other.id !== atom.id)
+                    .map((other) => `${other.element} (Line ${other.id + 1})`)
+                    .join(", ");
+
+                checks.push({
+                    type: "warning",
+                    message: `${atom.element} (Line ${atom.id + 1}) overlaps with ${otherAtoms}`,
```
Comment (VsevolodX) [inline src/material.js:341]: I'm thinking turning this into more generic message from an enum, to use like `messages.atomsOverlap` and define it all in a dedicated property.

File: src/basis/basis.js [hunk 2]
```diff
@@ -440,6 +440,42 @@ export class Basis {
...
+     */
+    getOverlappingAtoms() {
+        // to simplify calculations, convert to cartesian coordinates
+        const _basis = this;
+        _basis.toCartesian();
+        const { coordinates, elements } = _basis;
+        const overlaps = [];
+
+        coordinates.forEach((entry1) => {
+            coordinates.forEach((entry2) => {
```
Comment (timurbazhirov) [inline src/basis/basis.js:455]: We only need to loop over the remaining atoms, not all of them again

File: src/basis/basis.js [hunk 3]
```diff
@@ -440,6 +440,42 @@ export class Basis {
...
+                const tolerance =
+                    overlapCoefficient *
+                    (getElementAtomicRadius(el1) + getElementAtomicRadius(el2)); // in angstroms
+
+                const distance = math.vDist(entry1.value, entry2.value);
+                if (distance < tolerance) {
+                    overlaps.push({
+                        id1: entry1.id,
+                        id2: entry2.id,
+                        element1: elements.find((el) => el.id === entry1.id).value,
```
Comment (timurbazhirov) [inline src/basis/basis.js:469]: use el1 and el2

where is the schema for overlaps defined?

File: src/basis/basis.js [hunk 4]
```diff
@@ -440,6 +440,42 @@ export class Basis {
...
+        const _basis = this;
+        _basis.toCartesian();
+        const { coordinates, elements } = _basis;
+        const overlaps = [];
+
+        coordinates.forEach((entry1) => {
+            coordinates.forEach((entry2) => {
+                if (entry1.id === entry2.id) return; // Don't compare an atom with itself
+                const el1 = elements.find((el) => el.id === entry1.id).value;
+                const el2 = elements.find((el) => el.id === entry2.id).value;
```
Comment (timurbazhirov) [inline src/basis/basis.js:458]: Maybe use indes in forEach and  `getElementByIndex`?

File: src/material.js [hunk 5]
```diff
@@ -319,4 +319,44 @@ export class Material extends HasMetadataNamedDefaultableInMemoryEntity {
...
+
+        const basis = this.Basis;
+        const overlappingAtomsGroups = basis.getOverlappingAtoms();
+
+        overlappingAtomsGroups.forEach(({ id1, id2, element1, element2 }) => {
+            const key = `basis.coordinates.${id1}`;
+            keys.add(key);
+            messages.push({
+                key,
+                checkName: "atomsOverlap",
```
Comment (timurbazhirov) [inline src/material.js:354]: Why do we hardcode "atomsOverlap"?

File: src/basis/basis.js [hunk 6]
```diff
@@ -440,6 +440,42 @@ export class Basis {
         return math.precise(latticeSize, 4);
     }
 
+    /**
+     * @summary function returns an array of overlapping atoms within specified tolerance.
+     * @returns {{element: String, id: Number}[]}
+     */
+    getOverlappingAtoms() {
```
Comment (timurbazhirov) [inline src/basis/basis.js:447]: We should make this configurable and turn off when the number of atoms exceeds a certain value - i.e. 1000

File: src/material.js (hunk 5 above)
Comment (VsevolodX) [reply src/material.js]: Ideally, it should be something like `checkNamesEnum.material.basis.atomsOverlap` stored in some centralized repository. This sounds like definitions, core to the entities that get checked -- code.js ?
But for now not to overoptimize I left it like that.

**Pull request merged**